diffrax/_step_size_controller/jump_step_wrapper.py

```diff
@@ -0,0 +1,424 @@
+            rejected_buffer = None
+            i_reject = jnp.asarray(0)
+        else:
+            rejected_buffer = jnp.zeros(
+                (self.rejected_step_buffer_len,) + jnp.shape(t1), dtype=tdtype
+            )
+            # rejected_buffer[len(rejected_buffer)] = jnp.inf (see def of _get_t)
+            i_reject = jnp.asarray(self.rejected_step_buffer_len)
+
+        # Find index of first element of step_ts/jump_ts greater than t0
+        i_step = _find_index(t0, step_ts)
+        i_jump = _find_index(t0, jump_ts)
+        # Clip t1 to the next element of step_ts or jump_ts
+        t1, _ = _clip_ts(t0, t1, i_step, step_ts, False)
+        t1, jump_next_step = _clip_ts(t0, t1, i_jump, jump_ts, True)
+
+        state = _JumpStepState(
+            jump_next_step,
+            dt_proposal,
+            i_step,
+            i_jump,
+            i_reject,
+            rejected_buffer,
+            step_ts,
+            jump_ts,
+            inner_state,
+        )
+
+        return t1, state
+
+    def adapt_step_size(
+        self,
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: Y,
+        y1_candidate: Y,
+        args: Args,
+        y_error: Optional[Y],
+        error_order: RealScalarLike,
+        controller_state: _JumpStepState[_ControllerState],
+    ) -> tuple[
+        BoolScalarLike,
+        RealScalarLike,
+        RealScalarLike,
+        BoolScalarLike,
+        _JumpStepState[_ControllerState],
+        RESULTS,
+    ]:
+        (
+            made_jump,
+            prev_dt,
+            i_step,
+            i_jump,
+            i_reject,
+            rejected_buffer,
+            step_ts,
+            jump_ts,
+            inner_state,
+        ) = controller_state.get()
+
+        # Let the controller do its thing
+        (
+            keep_step,
+            next_t0,
+            next_t1,
+            _,
+            inner_state,
+            result,
+        ) = self.controller.adapt_step_size(
+            t0, t1, y0, y1_candidate, args, y_error, error_order, inner_state
+        )
+
+        # This is just a logging utility for testing purposes
+        if self.callback_on_reject is not None:
+            jax.debug.callback(self.callback_on_reject, keep_step, t1)
+
+        # Check whether we stepped over an element of step_ts/jump_ts/rejected_buffer
+        # This is all still bookkeeping for the PREVIOUS STEP.
+        if step_ts is not None:
+            # If we stepped to `t1 == step_ts[i_step]` and kept the step, then we
+            # increment i_step and move on to the next t in step_ts.
+            step_inc_cond = keep_step & (t1 == _get_t(i_step, step_ts))
+            i_step = jnp.where(step_inc_cond, i_step + 1, i_step)
+
+        if jump_ts is not None:
+            next_jump_t = _get_t(i_jump, jump_ts)
+            jump_inc_cond = keep_step & (t1 >= eqxi.prevbefore(next_jump_t))
+            i_jump = jnp.where(jump_inc_cond, i_jump + 1, i_jump)
```

Comment:
Likewise, if someone else is fiddling with the times then this seems to me like it might be fragile.

Recalling the previous implementation with its inefficient use of searchsorted. I think the robust approach here might be to write something with the same API as that, but whose implementation is just a simple linear search forwards or backwards from the current position (which is a 'hint' about where to start searching).

Most of the time that will just iterate once and be done, as here. But in the edge cases it should now do the right thing.